### PR TITLE
add initial implementation of CUSTOM action in AuthorizationPolicy

### PR DESCRIPTION
**Pull request by yangminzhu**
This PR implements the CUSTOM action and tested manually for now, remaining work:
1. Fix to use the namespace of MeshConfig
2. Fix to properly support k8s service and ServiceEntry
3. E2E tests with sample ext_authz server

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[x] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

Pull Request Attributes

Please check any characteristics that apply to this pull request. 

[x] Does not have any changes that may affect Istio users.

### CODE REVIEW

File: pilot/pkg/security/authz/builder/extauthz.go
Comment (incfly) [inline pilot/pkg/security/authz/builder/extauthz.go:71]: If seems like the `hasDuplicate` is acutally `notAllTheSame`, right? The function name is confusing.

Comment (incfly) [inline pilot/pkg/security/authz/builder/extauthz.go:71]: The condition is checking whether names are duplicated, but the error msg is about "only 1 allowed"?

File: pilot/pkg/security/authz/builder/extauthz.go
Comment (incfly) [inline pilot/pkg/security/authz/builder/extauthz.go:115]: Capitalizae log line

File: pilot/pkg/security/authz/builder/extauthz.go
Comment (incfly) [inline pilot/pkg/security/authz/builder/extauthz.go:185]: Is this comment outdated?

Comment (incfly) [reply pilot/pkg/security/authz/builder/extauthz.go:185]: Or you want to address in follow-up PR? If so, would need reword

File: pilot/pkg/security/authz/builder/extauthz.go
Comment (incfly) [inline pilot/pkg/security/authz/builder/extauthz.go:192]: Are there some model package util function for this kinda definition? I feel very likely

File: pilot/pkg/security/authz/builder/extauthz.go
Comment (incfly) [inline pilot/pkg/security/authz/builder/extauthz.go:218]: why it's required though? Just because of Envoy ext_authz config validation semantics? And this config does not take effect when we are using the Cluster generated by Istiod? If so, could you comment reason a bit?

File: pilot/pkg/security/authz/builder/extauthz.go
Comment (incfly) [inline pilot/pkg/security/authz/builder/extauthz.go:220]: Sounds like ext authz filter's wrong constraint as well, as above?

(no AI, just confirm)

File: pilot/pkg/security/authz/builder/extauthz.go
Comment (incfly) [inline pilot/pkg/security/authz/builder/extauthz.go:252]: If we replace the cluster name's character, does that require us to change the cds to duplicate the coresponding cluster and replace name with `-` as well? Or there're some magic with the replacing?

File: pkg/config/validation/extensionprovider.go
Comment (incfly) [inline pkg/config/validation/extensionprovider.go:46]: The func is passing a `status` string, without knowing it's actually passing the `x.status_on_error` field. the function naming seems is leaking details how how it should be used.

File: pilot/pkg/security/authz/builder/builder.go
Comment (incfly) [inline pilot/pkg/security/authz/builder/builder.go:221]: I read the sample output and code and here is my understanding: 

- we generate additional rbac filter, with shadow deny policies. Those rbac filter's condition is based on ext_authz condition.
- Then the ext authz filter is using the preivous rbac filte's output metadata filter, with `filterEnabledMetadata` to determine whether acutal kick in.

In some way, the first rbac is only to prepapre the triggering condition for the ext authz's config limitation?

Is that correct? Feel free to point to some section in the design doc. I do feel though, this is worth a bit more detailed comments around the code, maybe here and other places.

File: pilot/pkg/security/authz/builder/extauthz.go
Comment (incfly) [inline pilot/pkg/security/authz/builder/extauthz.go:40]: are you aware of any tooling(istioctl) or something is making dependency on the format of this rule's name? just seeing this is adding `-` in between in a slight different way than before.

Fine if it's totally internal.

File: pilot/pkg/networking/plugin/plugin.go
Comment (yangminzhu) [reply pilot/pkg/networking/plugin/plugin.go:23]: this is the common line argument in istiod, a global flag that controls the authn/z plugin itself, honestly I think we don't need this and should just enable authn/z plugins by default.

File: pilot/pkg/security/authz/builder/extauthz.go
Comment (yangminzhu) [reply pilot/pkg/security/authz/builder/extauthz.go:40]: this should be totally internal.

File: pilot/pkg/security/authz/builder/extauthz.go
Comment (yangminzhu) [reply pilot/pkg/security/authz/builder/extauthz.go:71]: ah, yes, updated the function name and reworded the error message.

Comment (yangminzhu) [reply pilot/pkg/security/authz/builder/extauthz.go:71]: see above.

File: pilot/pkg/security/authz/builder/extauthz.go
Comment (yangminzhu) [reply pilot/pkg/security/authz/builder/extauthz.go:185]: will be addressed in follow PR, updated comment.

File: pilot/pkg/security/authz/builder/extauthz.go
Comment (yangminzhu) [reply pilot/pkg/security/authz/builder/extauthz.go:192]: there is one in validation pkg but it caused some cyclic import issue.

File: pilot/pkg/security/authz/builder/extauthz.go
Comment (yangminzhu) [reply pilot/pkg/security/authz/builder/extauthz.go:220]: ext_authz filter is using referring to the generic HttpUri API, not only validation makes sense in the ext_authz context.

File: pilot/pkg/security/authz/builder/extauthz.go
Comment (yangminzhu) [reply pilot/pkg/security/authz/builder/extauthz.go:218]: required due to the validation semantic, the smaller timeout should take effect, updated comment.

File: pilot/pkg/security/authz/builder/extauthz.go
Comment (yangminzhu) [reply pilot/pkg/security/authz/builder/extauthz.go:252]: this is just replacing the authority used in the ext_authz config, the cluster name is not changed.

File: pkg/config/validation/extensionprovider.go
Comment (yangminzhu) [reply pkg/config/validation/extensionprovider.go:46]: I think it's fine, this is a local function (not exported) and is only supposed to be used in this package.

File: pilot/pkg/security/authz/builder/builder.go
Comment (yangminzhu) [reply pilot/pkg/security/authz/builder/builder.go:221]: that's correct, I think this is already pretty clear with the current comment, added a link to the design doc for more details.

**Comment by yangminzhu** [comment]
@incfly thanks for the review, updated PR to address comments so far.

Also @istio/wg-networking-maintainers-pilot can someone take a look from the networking side? thanks.

File: pilot/pkg/security/authz/builder/builder.go
Comment (liminw) [inline pilot/pkg/security/authz/builder/builder.go:122]: Why is custom builder using DENY filter?

**Comment by yangminzhu** [comment]
@costinm @howardjohn could you help take a look? thank you.

File: pilot/pkg/networking/plugin/authz/authorization.go
Comment (yangminzhu) [reply pilot/pkg/networking/plugin/authz/authorization.go:36]: it's used to distinguish the CUSTOM action from other actions (ALLOW/DENY/AUDIT), the CUSTOM action is handled separately and added to the beginning of the filter chain.

File: pilot/pkg/security/authz/builder/builder.go
Comment (yangminzhu) [reply pilot/pkg/security/authz/builder/builder.go:122]: DENY will cause an invalid config (e.g. found HTTP attribute in TCP filter chain) to always trigger the action (calling ext_authz), it doesn't matter in runtime as it's shadow mode (only evaluate but not enforce).

File: pilot/pkg/security/authz/builder/extauthz.go
Comment (incfly) [reply pilot/pkg/security/authz/builder/extauthz.go:221]: because envoy is validating as HTTP URI typed, i see.

"s/required/required for Envoy's validation/g" optional

File: pilot/pkg/security/authz/builder/builder.go
Comment (incfly) [reply pilot/pkg/security/authz/builder/builder.go:221]: Where's the link of the design doc? didn't see it nearby here.

We can add one sentence like "...keyed by the policy. And then the ext_authz filter can utilize these metadata to trigger the enforcement conditionally."

optional.

**Review by incfly — APPROVED** [review]

File: pilot/pkg/security/authz/builder/builder.go
Comment (liminw) [reply pilot/pkg/security/authz/builder/builder.go:122]: I see. Thanks.

**Review by liminw — APPROVED** [review]

File: pilot/pkg/security/authz/builder/builder.go
Comment (hzxuzhonghu) [inline pilot/pkg/security/authz/builder/builder.go:229]: Does this mean this filter will be executed twice?

File: pilot/pkg/security/authz/builder/extauthz.go
Comment (hzxuzhonghu) [inline pilot/pkg/security/authz/builder/extauthz.go:86]: This should be validated during mesh configuration loading

File: pilot/pkg/security/authz/builder/extauthz.go
Comment (hzxuzhonghu) [inline pilot/pkg/security/authz/builder/extauthz.go:92]: Can we cache the result

File: pilot/pkg/security/authz/builder/builder.go
Comment (yangminzhu) [reply pilot/pkg/security/authz/builder/builder.go:229]: It depends on your authz policy. If you have Authz policy with both CUSTOM and non-CUSTOM action, then yes. The 1st RBAC will only evaluate the matching rule. We already do this for DENY policy (it will be translated to a different RBAC filter).

File: pilot/pkg/security/authz/builder/extauthz.go
Comment (yangminzhu) [reply pilot/pkg/security/authz/builder/extauthz.go:86]: how? This needs the list of authz policies applied for a given workload, we don't know the workload in mesh configuration loading. Also I'm not sure how useful this could be, it seems the final result is mostly the same regardless where you do the validation.

Comment (yangminzhu) [reply pilot/pkg/security/authz/builder/extauthz.go:86]: oh, I looked at the wrong line (70), but actually this is also validated in mesh configuration loading in pkg/config/validation/extensionprovider.go

File: pilot/pkg/security/authz/builder/extauthz.go
Comment (yangminzhu) [reply pilot/pkg/security/authz/builder/extauthz.go:92]: probably in a follow PR, there might still some big changes to the implementation here, added a TODO.

File: pilot/pkg/security/authz/builder/builder.go
Comment (yangminzhu) [reply pilot/pkg/security/authz/builder/builder.go:221]: updated the comment, the design link is just below.

File: pilot/pkg/security/authz/builder/extauthz.go
Comment (hzxuzhonghu) [reply pilot/pkg/security/authz/builder/extauthz.go:86]: I am still confused about why we need to duplicate validating it.

File: pilot/pkg/security/authz/builder/extauthz.go
Comment (howardjohn) [inline pilot/pkg/security/authz/builder/extauthz.go:188]: do not use cluster.local, we support custom domains

File: pilot/pkg/security/authz/builder/extauthz.go
Comment (howardjohn) [inline pilot/pkg/security/authz/builder/extauthz.go:176]: Why is this namespace/service and not namespace/FQDN? doesn't that break ability to use ServiceEntry?

File: pilot/pkg/security/authz/builder/extauthz.go
Comment (howardjohn) [inline pilot/pkg/security/authz/builder/extauthz.go:219]: timeout in DR is connect timeout, not http request timeout

File: pilot/pkg/security/authz/builder/extauthz.go
Comment (howardjohn) [inline pilot/pkg/security/authz/builder/extauthz.go:256]: We already have a header compatible, BuildDNSSrvSubsetKey. Should we use this?

**Review by howardjohn — COMMENTED** [review]
LGTM aside from small comments

File: pilot/pkg/security/authz/builder/extauthz.go
Comment (yangminzhu) [reply pilot/pkg/security/authz/builder/extauthz.go:188]: yes, this is temporary and still under development (see comment above), I will fix this part in a follow up PR, it will touch more networking code and we will also likely get the hostname from the push context (e.g. serviceIndex).

File: pilot/pkg/security/authz/builder/extauthz.go
Comment (yangminzhu) [reply pilot/pkg/security/authz/builder/extauthz.go:256]: probably not, I think this is only used in `sni` but not the actual cluster name?

File: pilot/pkg/security/authz/builder/extauthz.go
Comment (yangminzhu) [reply pilot/pkg/security/authz/builder/extauthz.go:116]: AH, good point, updated.

File: pilot/pkg/security/authz/builder/extauthz.go
Comment (yangminzhu) [reply pilot/pkg/security/authz/builder/extauthz.go:100]: what do you mean `but ignore them`? We still return the errors if `errs != nil`. `extAuthz.http` should not be nil if `errs` is not nil, `extAuthz.tcp` might be but we will check it explicitly in builder.go.

File: pilot/pkg/security/authz/builder/extauthz.go
Comment (yangminzhu) [reply pilot/pkg/security/authz/builder/extauthz.go:86]: there are some validation that we can only do here because we don't know the workload and policies applied to it in ValidateMeshConfig().

File: pilot/pkg/security/authz/builder/extauthz.go
Comment (yangminzhu) [reply pilot/pkg/security/authz/builder/extauthz.go:219]: Good point, We currently don't expose this timeout in the API so we have to use some default value here, added a TODO to revisit this and investigate if we need to expose it later.

By the way, the http request timeout here is `measured since when the frame with end_stream flag is sent until when the first frame is received`. For ext_authz request (not stream) that is probably close to the connect timeout in DR?

File: pilot/pkg/security/authz/builder/extauthz.go
Comment (yangminzhu) [reply pilot/pkg/security/authz/builder/extauthz.go:176]: good question, I think we're trying to follow with the more general convention to refer to other (k8s or Istio) resource by using the name directly, and the indirection gives more flexibility (e.g. the user doesn't need to specify the cluster.local part)

This works well with k8s service but not with ServiceEntry because you can have multiple hosts. I think we will probably support only 1 host in ServiceEntry if it is being used in extension provider here. cc @louiscryan

**Comment by yangminzhu** [comment]
@howardjohn I addressed the comments so far, there are some issues that I have planned to fix in a upcoming PR (e.g. the cluster.local domain), would you mind to approve if this looks good to you now? thanks.

File: pilot/pkg/security/authz/builder/extauthz.go
Comment (howardjohn) [reply pilot/pkg/security/authz/builder/extauthz.go:259]: Why not? I don't want to invent a new format. BuildDNSSrvSubsetKey is meant for _exactly_ this use case.

File: pilot/pkg/security/authz/builder/extauthz.go
Comment (howardjohn) [reply pilot/pkg/security/authz/builder/extauthz.go:100]: ack, misread the code

**Review by howardjohn — APPROVED** [review]

File: pilot/pkg/security/authz/builder/extauthz.go
Comment (yangminzhu) [reply pilot/pkg/security/authz/builder/extauthz.go:259]: Do you mean to change the networking code to use BuildDNSSrvSubsetKey for the cluster name in the first place? If that is okay (not sure if it has any significant side effect),  that could be done as a separate refactor in the networking core code.

Here we are only referring to the already generated cluster name that still uses the `|` format.

Comment (howardjohn) [reply pilot/pkg/security/authz/builder/extauthz.go:259]: No, I mean instead of replacing `|` with `-` replace it with `_.` which is already the standardized format

Comment (yangminzhu) [reply pilot/pkg/security/authz/builder/extauthz.go:259]: Oh, I misunderstood your original comment, will do in a follow PR.

File: pilot/pkg/security/authz/builder/extauthz.go
Comment (hzxuzhonghu) [reply pilot/pkg/security/authz/builder/extauthz.go:86]: make sense

**Pull request merged**
